011-kafka-canary.md

```diff
@@ -0,0 +1,40 @@
+# Strmixi Canary
+
+Implement a Kafka canary tool which will act as an indicator of whether Kafka clusters are operating correctly. This would be achieved by creating a canary topic and continously producing and consuming events on the topics. 
+
+## Current situation
+
+Currently  Strimzi does not have any canary style feature. It does have readiness checks for Kafka cluster which can be replaced/augmented with the canary.
+
+## Motivation
+
+The canary would provide the following:
+  
+  * Improved readiness check for Kafka clusters
+  * SLI(s) for service admins
+  * Metrics:
+    * End-to-end latency
+    * Succesfull/failed events
+    * Disk usage
+    * Disk errors
+
+## Proposal
+
+The canary would be implemented as part of the Kafka Roller component. 
+A single _canary_ topic would be created on creation of a Kafka cluster. The topic would have one partition for Kafka broker. Events would then be continously produced and consumed to/from the canary topic. The internal TLS 9091 port would be used to connect to the brokers.
+
+As this happens the relevant metrics would be collected and exported. Errors such as the inability to produce to the topic can then be used by the KafkaRoller to trigger a restart, acting as healthc check for the Kafka cluster. 
```

Comment:
I think this needs significantly more details on how it will be implemented. Will it be a separate pod deployed by the operator? How will it pass the results to the KafkaRoller? Or will it run inside the operator? If yes, how it will be implemented there with respect to resource consumptions, working in parallel etc.?

It should probably also go into the details how would the canary create and configure its topic and evaluate the result. Would it have topic with replication factor (RF) == min.isr == number of Kafka nodes? If yes, how will it deal with rolling updates etc.? Or will the min.isr be RF-1? How would such topic change the replication factor during scale-ups/scale-downs? If it will have less replicas then nodes, how will it test the whole cluster?

Reply:
Adding more resources consumption to the cluster operator was my first concern but even talking with @tombentley it was clear that we are moving resources consumption somewhere else if the canary is running in a different pod with another JVM and we also need to expose a kind of REST (?) interface to the Kafka Roller to get canary result.
Anyway, we should try to narrow down the resource usage for producer and consumer (buffer sizes, ...) and even the partitions size (maybe even using a compacted topic).

As I mentioned in another comment above what's the value of having RF > 0 if the objective of the canary tool is checking that a specific broker is working or not, when it's hosting one leader partition that would not have sense having a different leader on a different broker because it's already hosting its own "canary" partition. What am I missing?

Reply:
In the first place, this needs to be described in the proposal. There are use cases where the CO operates many clusters. In that case running all of this in single pod will not be nice despite the possible memory savings. Especially if you would have it running constantly and sending a message every second or so. So I think a separate pod deployed as part of the cluster would be better here.